### PR TITLE
[FLINK-28510][python][connector] Support using new KafkaSink API

### PR DESCRIPTION
**Pull request by Vancior**
## What is the purpose of the change

This PR introduces KafkaSink API in PyFlink, aligning the Java new KafkaSink API.

## Brief change log

- add KafkaSink, KafkaSinkBuilder, KafkaRecordSerializationSchema, KafkaRecordSerializationSchemaBuilder API
- add optional transformation for Sink, to support selecting the sink topic for each record as additional row info in Python and then do selection in Java KafkaSink

## Verifying this change

This change added tests and can be verified as follows:

- `KafkaSinkTests` in test_kafka.py

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs & Python Sphinx doc)

### CODE REVIEW

File: flink-python/pyflink/common/serialization.py [hunk 1]
```diff
@@ -38,6 +40,10 @@ class SerializationSchema(object):
     def __init__(self, j_serialization_schema=None):
         self._j_serialization_schema = j_serialization_schema
 
+    @abstractmethod
+    def require_row_type(self) -> bool:
```
Comment (dianfu) [inline flink-python/pyflink/common/serialization.py:44]: Seems a little wired, what about removing this method?

File: flink-python/pyflink/datastream/connectors/base.py [hunk 2]
```diff
@@ -52,6 +53,24 @@ def __init__(self, sink: Union[str, JavaObject]):
...
 
 
+class TransformAppender(ABC):
+
+    @abstractmethod
+    def apply(self, ds):
+        pass
+
+
+class PreTransformWrapper(ABC):
```
Comment (dianfu) [inline flink-python/pyflink/datastream/connectors/base.py:63]: Rename it to SupportsPreprocessing?

File: flink-python/pyflink/datastream/connectors/base.py [hunk 3]
```diff
@@ -52,6 +53,24 @@ def __init__(self, sink: Union[str, JavaObject]):
...
+
+    @abstractmethod
+    def need_pre_transform(self) -> bool:
+        pass
+
+    @abstractmethod
+    def get_pre_transform(self) -> 'TransformAppender':
```
Comment (dianfu) [inline flink-python/pyflink/datastream/connectors/base.py:70]: apply(self, ds) -> 'ds'?

File: flink-python/pyflink/datastream/connectors/base.py [hunk 4]
```diff
@@ -52,6 +53,24 @@ def __init__(self, sink: Union[str, JavaObject]):
...
+
+    @abstractmethod
+    def apply(self, ds):
+        pass
+
+
+class PreTransformWrapper(ABC):
+
+    @abstractmethod
+    def need_pre_transform(self) -> bool:
```
Comment (dianfu) [inline flink-python/pyflink/datastream/connectors/base.py:66]: remove this method?

File: flink-python/src/main/java/org/apache/flink/python/util/PythonConnectorUtils.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /**
+     * A {@link SerializationSchema} for {@link Row} that only serialize the second column using a
+     * wrapped {@link SerializationSchema} for {@link T}.
+     *
+     * @param <T> The actual data type wrapped in the Row.
+     */
+    public static class SecondColumnSerializationSchema<T> implements SerializationSchema<Row> {
+
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonConnectorUtils.java:71]: Add `private static final long serialVersionUID = 1L`;

File: flink-python/src/main/java/org/apache/flink/python/util/PythonConnectorUtils.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    public static <T> T createFirstColumnSelector(Class<T> clazz) {
+        return (T)
+                Proxy.newProxyInstance(
+                        clazz.getClassLoader(),
+                        new Class[] {clazz},
+                        new FirstColumnSelectorInvocationHandler());
+    }
+
+    /** The serializable {@link InvocationHandler} as the proxy for first column selector. */
+    public static class FirstColumnSelectorInvocationHandler
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonConnectorUtils.java:50]:
```suggestion
    public static class FirstColumnTopicSelectorInvocationHandler
```

File: flink-python/pyflink/datastream/connectors/base.py (hunk 4 above)
Comment (Vancior) [reply flink-python/pyflink/datastream/connectors/base.py]: This is still needed that a sink implementing `SupportsPreprocessing` may not require preprocessing, e.g. `KafkaSink` doesn't need preprocessing when using fixed topic.

File: flink-python/pyflink/datastream/connectors/kafka.py [hunk 7]
```diff
@@ -816,3 +826,340 @@ def offsets(offsets: Dict['KafkaTopicPartition', int],
...
+        self._preprocessing = None
+
+    def build(self) -> 'KafkaSink':
+        """
+        Constructs the :class:`KafkaSink` with the configured properties.
+        """
+        if self._preprocessing is None:
+            return KafkaSink(self._j_builder.build())
+        else:
+            return KafkaSink(self._j_builder.build(), self._preprocessing)
```
Comment (bgeng777) [inline flink-python/pyflink/datastream/connectors/kafka.py:912]: directly `return KafkaSink(self._j_builder.build(), self._preprocessing)` ?

**Review by HuangXingBo — APPROVED** [review]

**Pull request closed without merging**